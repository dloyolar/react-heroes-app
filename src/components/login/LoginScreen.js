import { useNavigate } from 'react-router-dom';
import { useContext } from 'react';
import { AuthContext } from '../../auth/authContext';
import { types } from '../../types/types';

export const LoginScreen = () => {
  const navigate = useNavigate();
  const { dispatch } = useContext(AuthContext);

  const handleLogin = () => {
    const action = {
      type: types.login,
      payload: { name: 'Diego' },
    };
    dispatch(action);

    const lastPath = localStorage.getItem('lastPath') || '/marvel';

    navigate(lastPath, {
      replace: true,
    });
  };

  return (
    <div className="container mt-5">
      <h1>Login</h1>
      <p className="text-muted">
        (Don't worry, this is a fake login, you dont need any user or
        credentials ðŸ‘Œ)
      </p>
      <hr />
      <button className="btn btn-primary" onClick={handleLogin}>
        Login
      </button>
    </div>
  );
};
