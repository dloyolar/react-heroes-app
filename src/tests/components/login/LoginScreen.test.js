import { mount } from 'enzyme';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import { LoginScreen } from '../../../components/login/LoginScreen';
import { types } from '../../../types/types';
import { AuthContext } from '../../../auth/authContext';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));

describe('Test on <LoginScreen />', () => {
  const contextValue = {
    dispatch: jest.fn(),
    user: {
      logged: false,
    },
  };
  const wrapper = mount(
    <AuthContext.Provider value={contextValue}>
      <MemoryRouter initialEntries={['/login']}>
        <Routes>
          <Route path="/login" element={<LoginScreen />} />
        </Routes>
      </MemoryRouter>
    </AuthContext.Provider>
  );
  test('should be display correctly and match with snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should dispatch have been called and navigate', () => {
    const handleClick = wrapper.find('button').prop('onClick');
    handleClick();
    expect(contextValue.dispatch).toHaveBeenCalledWith({
      payload: { name: 'Diego' },
      type: types.login,
    });

    expect(mockNavigate).toHaveBeenCalledWith('/marvel', { replace: true });

    localStorage.setItem('lastPath', '/dc');

    handleClick();

    expect(mockNavigate).toHaveBeenCalledWith('/dc', { replace: true });
  });
});
