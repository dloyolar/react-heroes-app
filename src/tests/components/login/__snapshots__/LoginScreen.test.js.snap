// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <LoginScreen /> should be display correctly and match with snapshot 1`] = `
<div
  className="container mt-5"
>
  <h1>
    Login
  </h1>
  <p
    className="text-muted"
  >
    (Don't worry, this is a fake login, you dont need any user or credentials ðŸ‘Œ)
  </p>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
